rust/rbac-registration/src/registration/cardano/mod.rs

```diff
@@ -40,32 +40,146 @@ pub struct RegistrationChain {
 }
 
 impl RegistrationChain {
+    /// Attempts to initialize a new RBAC registration chain
+    /// from a given CIP-509 registration, ensuring uniqueness of Catalyst ID, stake
+    /// addresses, and associated public keys.
+    ///
+    /// # Errors
+    ///  - Propagates any I/O or provider-level errors encountered while checking key
+    ///   ownership (e.g., database lookup failures).
+    pub async fn new<State>(
+        cip509: &Cip509,
+        state: &mut State,
+    ) -> anyhow::Result<Option<Self>>
+    where
+        State: RBACState,
+    {
+        let Some(new_chain) = Self::new_stateless(cip509) else {
+            return Ok(None);
+        };
+
+        // Verify that a Catalyst ID of this chain is unique.
+        {
+            let cat_id = new_chain.catalyst_id();
+            if state.is_chain_known(cat_id).await? {
+                cip509.report().functional_validation(
+            &format!("{} is already used", cat_id.as_short_id()),
+                "It isn't allowed to use same Catalyst ID (certificate subject public key) in multiple registration chains",
+                    );
+            }
+
+            // Checks that a new registration doesn't contain a signing key that was used by any
+            // other chain. Returns a list of public keys in the registration.
+            for role in cip509.all_roles() {
+                if let Some(key) = cip509.signing_pk_for_role(role) {
+                    if let Some(previous) = state.chain_catalyst_id_from_public_key(&key).await? {
+                        if &previous != cat_id {
+                            cip509.report().functional_validation(
+                                &format!("An update to {cat_id} registration chain uses the same public key ({key:?}) as {previous} chain"),
+                                "It isn't allowed to use role 0 signing (certificate subject public) key in different chains",
+                            );
+                        }
+                    }
+                }
+            }
```

Comment:
As far as I can see, the exact same code is used in the `update` function. Can we reuse it?